pkg/admission/validator/lakom/lakom_suite_test.go

```diff
@@ -0,0 +1,130 @@
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+func TestLakom(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Lakom Suite")
+}
+
+var _ = Describe("Shoot validator", func() {
+	var (
+		ctx = context.Background()
+
+		shootValidator extensionswebhook.Validator
+		ctrl           *gomock.Controller
+		apiReader      *mockclient.MockReader
+
+		shoot *core.Shoot
+	)
+
+	Describe("#Validate", func() {
+		BeforeEach(func() {
+			scheme := runtime.NewScheme()
+			utilruntime.Must(apilakom.AddToScheme(scheme))
+			utilruntime.Must(v1alpha1.AddToScheme(scheme))
+
+			decoder := serializer.NewCodecFactory(scheme, serializer.EnableStrict).UniversalDecoder()
+			apiReader = mockclient.NewMockReader(ctrl)
+			ctrl = gomock.NewController(GinkgoT())
+
+			shootValidator = lakom.NewShootValidator(apiReader, decoder)
+
+			shoot = &core.Shoot{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "garden-tst",
+					Name:      "tst",
+				},
+				Spec: core.ShootSpec{
+					Extensions: []core.Extension{
+						{
+							Type: "shoot-lakom-service",
+							ProviderConfig: &runtime.RawExtension{
+								Raw: encode(&v1alpha1.LakomConfig{
+									TypeMeta: metav1.TypeMeta{
+										APIVersion: v1alpha1.SchemeGroupVersion.String(),
+										Kind:       "LakomConfig",
+									},
+									Scope: apilakom.Cluster,
+								}),
+							},
+						},
+					},
+				},
+			}
+		})
+
+		It("should return err when new is not a Shoot", func() {
+			err := shootValidator.Validate(ctx, &corev1.Pod{}, nil)
+
+			Expect(err).To(HaveOccurred())
+		})
+
+		It("should do nothing when the Shoot does no specify a shoot-lakom-service extension", func() {
+			shoot.Spec.Extensions[0].Type = "foo"
+
+			Expect(shootValidator.Validate(ctx, shoot, nil)).To(Succeed())
+		})
+
+		It("should return err when shoot-lakom-service providerConfig cannot be decoded", func() {
+			shoot.Spec.Extensions[0].ProviderConfig = &runtime.RawExtension{
+				Raw: []byte(`{"bar": "baz"}`),
+			}
+
+			err := shootValidator.Validate(ctx, shoot, nil)
+			Expect(err).To(HaveOccurred())
+		})
+
+		It("should succeed for valid Shoot", func() {
+			Expect(shootValidator.Validate(ctx, shoot, nil)).To(Succeed())
+		})
+
+		It("should fail if the given scope is not recognized", func() {
+			extension := &runtime.RawExtension{
+				Raw: encode(&v1alpha1.LakomConfig{
+					TypeMeta: metav1.TypeMeta{
+						APIVersion: v1alpha1.SchemeGroupVersion.String(),
+						Kind:       "LakomConfig",
+					},
+					Scope: "invalid",
+				}),
+			}
+
+			shoot.Spec.Extensions[0].ProviderConfig = extension
+
+			err := shootValidator.Validate(ctx, shoot, nil)
+			Expect(err).To(ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{
+				"Type":  Equal(field.ErrorTypeInvalid),
+				"Field": Equal("spec.extensions[0].providerConfig.scope"),
+			}))))
+		})
+
+	})
+})
+
+func encode(obj runtime.Object) []byte {
+	data, _ := json.Marshal(obj)
```

Comment:
This error should not be ignored. Maybe you can avoid using the `encode` function by setting Raw directly to the byte array representation of a string 

```go
	extension := &runtime.RawExtension{
		Raw: []byte(`apiVersion: lakom.extensions.gardener.cloud/v1alpha1
kind: LakomConfig
scope: invalid
`),
```

Reply:
I do like that with this encode we can leverage the type system to avoid stupid mistakes with spelling at compile time. I'd prefer if we just handle the error. This error isn't something that we're testing and this code being just a unit test, are you okay with just panicking if there's an error?

Reply:
You can also add check that the error have not occurred here

```suggestion
	data, err := json.Marshal(obj)
	Expect(err).ToNot(HaveOccurred())
```

Reply:
I try to avoid testing for things that are outside the scope of the unit we're writing the tests for. Especially if the only way I see this could fail is if there's a bug somewhere in the `json` lib, in its dependencies, or we've passed something malformed somehow.

Either way, I don't have a super strong opinion on this and I'm interested to see if you have some opinions on this topic.

Reply:
We are more interested that the `obj` could be successfully marshaled, it is not expected to catch bugs in json.Marshal. Just ignoring the error is not correct because then we cannot rely the value in `data` variable is correct. 

I am not sure how panic will be helpful here, I guess it could interrupt the execution of the other tests and might not provide meaningful/helpful information what was the input when the panic occurred. 

To sum up, I am fine just to ensure that there `json.Marshal` does not return error.